### PR TITLE
Connect outage vars to powerflow

### PR DESCRIPTION
**Pull request by brtietz**
Adds crit_load_escalation to the relevant cmods. Brings the variables from the vartables into battery->advance such that the powerflow variables can use the outage code.

Needs to be paired with a SAM PR based on critical-load-ui branch, @cpaulgilman to create.

A future PR (ssc branch dispatch_for_grid_outage) will implement the outage behavior in dispatch, and clean up edge cases like the inverter night time loads.

Via email we had proposed making grid_outage and crit_load SSC outputs - should that happen on this PR or a different PR? If this one, which cmods are responsible for the output?

### CODE REVIEW

File: ssc/cmod_battery.cpp [hunk 1]
```diff
@@ -2160,11 +2187,11 @@ class cm_battery : public compute_module
 
                         if (resilience) {
                             resilience->add_battery_at_outage_timestep(*batt->dispatch_model, lifetime_idx);
-                            resilience->run_surviving_batteries(p_crit_load[lifetime_idx % n_rec_single_year], power_input_lifetime[lifetime_idx]);
+                            resilience->run_surviving_batteries(p_crit_load_full[lifetime_idx], power_input_lifetime[lifetime_idx]);
```
Comment (sjanzou) [inline ssc/cmod_battery.cpp:2190]: Is p_crit_load_full the percentage (or timeseries) entered by the user?

File: ssc/cmod_battery.cpp [hunk 2]
```diff
@@ -2110,8 +2119,26 @@ class cm_battery : public compute_module
                         throw exec_error("battery", "If run_resiliency_calcs is 1, crit_load must have length > 0 and values > 0");
                     }
                 }
+                if (!crit_load_specified && batt->analyze_outage) {
+                    throw exec_error("battery", "If grid_outage is specified in any time step, crit_load must have length > 0 and values > 0");
```
Comment (sjanzou) [inline ssc/cmod_battery.cpp:2123]: Thanks for adding this!

File: ssc/cmod_pvsamv1.cpp [hunk 3]
```diff
@@ -2215,12 +2238,12 @@ void cm_pvsamv1::exec()
 
                 if (resilience) {
                     resilience->add_battery_at_outage_timestep(*batt->dispatch_model, idx);
-                    resilience->run_surviving_batteries(p_crit_load_in[idx % nrec], sharedInverter->powerAC_kW, dcPower_kW,
+                    resilience->run_surviving_batteries(p_crit_load_full[idx], sharedInverter->powerAC_kW, dcPower_kW,
```
Comment (sjanzou) [inline ssc/cmod_pvsamv1.cpp:2241]: All are full lifetime now, yeah!

**Comment by sjanzou** [comment]
> Adds crit_load_escalation to the relevant cmods. Brings the variables from the vartables into battery->advance such that the powerflow variables can use the outage code.
> 
> Needs to be paired with a SAM PR based on critical-load-ui branch, @cpaulgilman to create.
> 
> A future PR (ssc branch dispatch_for_grid_outage) will implement the outage behavior in dispatch, and clean up edge cases like the inverter night time loads.
> 
> Via email we had proposed making grid_outage and crit_load SSC outputs - should that happen on this PR or a different PR? If this one, which cmods are responsible for the output?

I am branching off of your branch and Paul's SAM branch and including the monthly outputs for the Puerto Rico project

**Review by sjanzou — APPROVED** [review]
Looks good, I am running into a hanging case at 67.8% file attached to this pull request
image

SAM project file case "PV Com grid out hanging" 
[Test_Case.zip](https://github.com/NREL/ssc/files/7241375/Test_Case.zip)

**Comment by sjanzou** [comment]
> Looks good, I am running into a hanging case at 67.8% file attached to this pull request
> image
> 
> SAM project file case "PV Com grid out hanging"
> [Test_Case.zip](https://github.com/NREL/ssc/files/7241375/Test_Case.zip)

The issue is with peak shaving dispatch and not the grid outage. With either case in the above SAM file, the peak dispatch option will hang in SAM at 67/8% and debugging in the SDKTool, after the fixes in the Fix_issues branches of ssc and SAM to generate the lk script and match the grid targets dispatch option results, the issue is a stack overflow in the BatteryPowerFlow::calculateACConnected:
image

**Comment by sjanzou** [comment]
Next issue is hanging for grid power targets and timestep resilency
image
image

Same issue as above with Test_Case.sam second case with time step resilency
image

image

Trace issue to possible incorrect sign - above cases run fine when sign is changed per pull request 619
image

File: ssc/cmod_battery.cpp (hunk 1 above)
Comment (brtietz) [reply ssc/cmod_battery.cpp:2190]: `p_crit_load_full` is a timeseries extended to lifetime by the compute module. `p_crit_load` is the year 1 timeseries. The GUI converts percentages into a kW timeseries in the "$MIMO$ elecload" equations on the Electric Load Crit page.

**Comment by sjanzou** [comment]
#620 merge still fails for case 3 of attached file per comment
[Test_Case.zip](https://github.com/NREL/ssc/files/7247377/Test_Case.zip)

**Comment by sjanzou** [comment]
Can we drop #618 and just use 620 and the fix you do in dispatch_for_grid_outage?

File: shared/lib_battery_powerflow.cpp [hunk 4]
```diff
@@ -469,7 +469,7 @@ void BatteryPowerFlow::calculateACConnected()
 
     // Error checking trying to charge from grid when not allowed
     if (!m_BatteryPower->canGridCharge && P_battery_ac < -tolerance) {
-        if ((fabs(P_grid_ac - P_grid_to_load_ac) > tolerance) && (-P_grid_ac > P_grid_to_load_ac)) {
+        if ((fabs(P_grid_ac - P_grid_to_load_ac) > tolerance) && (-P_grid_ac > P_grid_to_load_ac + tolerance)) {
```
Comment (sjanzou) [inline shared/lib_battery_powerflow.cpp:472]: Is this in 620, too?

**Review by sjanzou — CHANGES_REQUESTED** [review]
Can we drop #618 and just use 620 and the fix you do in dispatch_for_grid_outage? Or, merge both in and then start a new branch to fix the resilency runner issue?

**Comment by cpaulgilman** [comment]
This SAM pull request should be ready to merge with this when ready: https://github.com/NREL/SAM/pull/707

File: shared/lib_battery_powerflow.cpp (hunk 4 above)
Comment (brtietz) [reply shared/lib_battery_powerflow.cpp:472]: I merged 620 into this branch, so yes.

**Comment by brtietz** [comment]
@sjanzou  In the interest of getting this merged sooner, I brought over a few key lines from `dispatch_for_grid_outages` that I think solve the stack overflow problems on this branch. The third test case works for me, let me know if anything else breaks.

`dispatch_for_grid_outage` is still broken as of this writing, since that code introduced some new problems. I'll try to get those problems fixed this week.

**Review by sjanzou — APPROVED** [review]
All Test_case.sam cases running now with this branch (dispatch_for_grid_outages still broken as noted), this looks good to merge along with corresponding SAM pull request from Paul. Thanks!

**Pull request merged**